lib/spack/spack/util/web.py

```diff
@@ -299,7 +299,7 @@ def find_versions_of_archive(archive_urls, list_url=None, list_depth=0):
     # Add '/' to the end of the URL. Some web servers require this.
     additional_list_urls = set()
     for lurl in list_urls:
-        if not lurl.endswith('/'):
+        if not lurl.endswith('/') and not lurl.endswith('.html'):
```

Comment:
I can't decide how I feel about this.  It makes me twitch as a former web developer, as some servers are going to have `.html`, windows ones can have `.htm`, and technically anything with an extension is likely to be a file.

Here are some corner cases I think might still be broken by #8429 after this change:

```
$ git grep list_url | egrep '\.[^/=]*"' | grep -v 'html"'
packages/eospac/package.py:    list_url = "https://laws.lanl.gov/projects/data/eos/eospacReleases.php"
packages/r-data-table/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/data.table"
packages/r-fit-models/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/fit.models"
packages/r-futile-logger/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/futile.logger"
packages/r-futile-options/package.py:    list_url = "https://cran.rstudio.com/src/contrib/Archive/futile.options"
packages/r-lambda-r/package.py:    list_url = "https://cran.rstudio.com/src/contrib/Archive/lambda.r"
packages/r-matr/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/matR/matR_0.9.tar.gz"
packages/r-methodss3/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/R.methodsS3"
packages/r-oo/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/R.oo"
packages/r-processx/package.py:    list_url      = "https://cran.r-project.org/src/contrib/Archive/processx/processx_3.1.0.tar.gz"
packages/r-ps/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/ps/ps_1.0.0.tar.gz"
packages/r-readxl/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/readxl/readxl_1.0.0.tar.gz"
packages/r-rpart-plot/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/rpart.plot"
packages/r-statnet-common/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/statnet.common"
packages/r-th-data/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/TH.data"
packages/r-utils/package.py:    list_url = "https://cran.r-project.org/src/contrib/Archive/R.utils"
```

Note that I grepped for anything with a `.` and an extension that is *not* `html`.  Are these cases broken too?

Given that it doesn't hurt to add more candidates to `additional_list_urls`, why not add *both* `<url>` and `<url>/` either:
1. in all cases, or
2. just when there is a `.` in `os.path.basename(url)`

I can't decide which is better.  Thoughts?

Reply:
> This makes me twitch as a former web developer, as some servers are going to have .html, windows ones can have .htm, and technically anything with an extension is likely to be a file.

Agreed, at what point should we use a better URL library like `requests`?

> Here are some corner cases I think might still be broken by #8429 after this change:

The `eospac` one might be broken, but we can't parse the URLs anyway so it doesn't matter. The CRAN ones are all non-broken, as those are directories and the package names happen to contain a period.

> Given that it doesn't hurt to add more candidates to additional_list_urls, why not add both <url> and <url>/ either:

Previously, we were adding both (if the url did not end in a slash). This was what caused #11680. I'm not sure why this is the case, but adding more list_urls doesn't help.

I'm fine with adding `.htm` and `.php` if you want, but I don't know of an easy way to handle this situation.

Reply:
Does requests actually handle this situation?

Reply:
No idea, I've never used it before. I just know it's popular.